### PR TITLE
mon: have mon-specific features & rework internal monmap structures

### PR DESCRIPTION
**Pull request by jecluis**
This pull request is the resulting branch from merging two features:

1. mon-specific features (previously, PR #7556)
2. holding mon addr and friends on a structure dedicated to hold mon info (previously, PR #6855)

Further info on each will be found at their original PR's.

`Signed-off-by: Joao Eduardo Luis <joao@suse.de>`

### CODE REVIEW

**Comment by jecluis** [comment]
fwiw, the whole thing has gone through a rados thrash suite. Failed tests have been rerun successfully.

File: src/mon/MonMap.h [hunk 1]
```diff
@@ -19,44 +19,176 @@
...
+    // Please note that we use a 'set' here instead of resorting to
+    // std::sort() because we need more info than that's available in
+    // the vector. The vector will thus be ordered by, e.g., public_addr
+    // while only containing the names of each individual monitor.
+    // The only way of achieving this with std::sort() would be to first
+    // insert every mon_info_t entry into a vector 'foo', std::sort() 'foo'
+    // with custom comparison functions, and then copy each invidual entry
+    // to a new vector. Unless there's a simpler way, we don't think the
+    // added complexity makes up for the additional memory usage of a 'set'.
+    set<mon_info_t, rank_cmp> tmp;
```
Comment (tchaikov) [inline src/mon/MonMap.h]: - can we move the implementation of this method into MonMap.cc to speed up the compilation?
- could also define the comparator using `WRITE_CMP_OPERATORS_2`.

File: src/mon/MonMap.h [hunk 2]
```diff
@@ -64,116 +196,154 @@ class MonMap {
   uuid_d& get_fsid() { return fsid; }
 
   unsigned size() {
-    return mon_addr.size();
+    return mon_info.size();
```
Comment (tchaikov) [inline src/mon/MonMap.h]: could take this chance to mark `size()` as a const method.

File: src/mon/MonMap.h [hunk 3]
```diff
@@ -64,116 +196,154 @@ class MonMap {
...
+  /**
+   * Check if monmap contains a monitor with address @p a
+   *
+   * @note checks for all addresses a monitor may have, public or otherwise.
+   *
+   * @param a monitor address
+   * @returns true if monmap contains a monitor with address @p;
+   *          false otherwise.
+   */
   bool contains(const entity_addr_t &a) {
```
Comment (tchaikov) [inline src/mon/MonMap.h]: could add `const` here also.

File: src/mon/MonmapMonitor.h [hunk 4]
```diff
@@ -83,6 +84,26 @@ class MonmapMonitor : public PaxosService {
 
  private:
   bufferlist monmap_bl;
+
+  class C_ApplyFeatures : public Context {
```
Comment (tchaikov) [inline src/mon/MonmapMonitor.h]: as this class is not used elsewhere, can we define it in MonmapMonitor.cc at the place where it is used, so we are able to find it's implementation right where it is used?

File: src/mon/mon_types.h [hunk 5]
```diff
@@ -234,4 +235,217 @@ struct C_MonOp : public Context
...
+  bool empty() const {
+    return features == 0;
+  }
+
+  /**
+   * Set difference of our features in respect to @p other
+   *
+   * Returns all the elements in our features that are not in @p other
+   *
+   * @returns all the features in @p other but not in our features.
```
Comment (tchaikov) [inline src/mon/mon_types.h]: this line conflicts with the line above. but i love the above one better =)

**Comment by liewegas** [comment]
@jecluis ping

**Comment by liewegas** [comment]
I think this has to wait.  And I can't think of a reason why having this field in jewel helps us...

**Comment by jecluis** [comment]
Yeah, it can wait as there's nothing actually leveraging it. It would just have been nice in order to time it with the jewel feature bit, but assuming we'll have a kraken feature bit it shouldn't matter much.

File: src/mon/MonMap.h (hunk 1 above)
Comment (jecluis) [inline src/mon/MonMap.h]: is it just me thinking that `WRITE_CMP_OPERATORS_2`, along with all the remaining macros for more than one parameter, sort of just obfuscate the code?

Comment (tchaikov) [inline src/mon/MonMap.h]: right. agree with you. cross that out!

File: src/mon/mon_types.h (hunk 5 above)
Comment (jecluis) [inline src/mon/mon_types.h]: yes, this was an artifact from before the semantic change. it'll be fixed on the next push.

**Comment by jecluis** [comment]
see https://github.com/ceph/ceph/pull/10907 instead

**Pull request closed without merging**